### DISCUSSION TITLE
Adding Locking / Confirmations / Log Info to Authentic

### DISCUSSION THREAD

**Discussion by stephendolan** [luckyframework/lucky#1285] [Feature requests & Ideas]
We have the following roadmap item:

> Authentic confirmable/lockable/log_info. Not too hard to do yourself, but a lot of people want this out-of-the-box. Not sure it’s that important but would be cool

Having just recently added email confirmations to Lucky Jumpstart [here](https://github.com/stephendolan/lucky_jumpstart/pull/106), I actually think this is worth a discussion about how much Lucky can/should take on.

I thought about using Teeplate for this in our generators, but that seems like it would get tedious for all of the various permutations of features you might want. Are these kinds of application features something we should consider building a more generic template capability like [what Rails has](https://guides.rubyonrails.org/rails_application_templates.html)? Otherwise, I think we'll constantly be pulled by the community towards implementing a massive set of base options that folks will only end up using a small portion of for most applications.

**Comment by jwoertink** [comment]
The Lucky philosophy is geared more toward "batteries included" with the option of escape hatch for you to do your own thing. I think it makes sense for Authentic to have these options built-in. Our generators would just say "Do you want auth", and that's it. We wouldn't customize it for you. Once you have auth, if you wanted locking or confirmations, that would be up to you to look up the docs on how to enable it.

I'm not a fan of the rails template stuff. I would like to stay away from going that route. We can always point people toward using that bloat gem and setting up their own stuff for that sort of thing.

**Reply by stephendolan** [reply]
I love the idea of implementing these specifically, and even going the Devise route where you basically just include the mixins that you want. If we did user logging like the roadmap item suggests, we'd probably want to keep GDPR in mind as well.

I'm curious though, what is it about the Rails app templates that you don't like specifically? From a user perspective I've always found them to be super productive for "building on the shoulders of giants" that have already figured out a bunch of this stuff. Additionally, I loved being able to run a template and then customize/rename stuff before committing since it's just my application code, not something I have to jump into a library to override with method overloading or config values.

**Reply by jwoertink** [reply]
I've always found that the templates get me like 60% of the way there, but leave me in a position of not understanding what all is in my app. Harder when in a team and everyone has to agree on a specific template that works for what we need.

I think a template system is super nice on quick client projects. The kind where you can quickly mass produce apps for a cheap turn and burn type deal. At that point I'd want to scaffold quickly as much as I can to move on to the next site. But I don't start new projects near enough for it to make sense.

**Reply by jwoertink** [reply]
You know, thinking about it... We do have the `init.custom` already. I guess if enough people wanted this feature of a quick template system, we could add a flag to that, but leave it off the main wizard. 🤔

**Reply by stephendolan** [reply]
Oh yeah, 100% agreed that this would mostly be leveraged for new folks coming to the framework. 

Honestly, part of me wanting something like this is motivated by not wanting the core team to have to bike shed for hours when someone opens up a discussion like "Switch Lucky default apps to TypeScript". I'm sure we'd all have valid opinions in either direction, but there's a sort of freedom in "Well, we ship with JS but if you want to submit a PR to add an option in the generator that's cool". 

Same goes with CSS frameworks. IMO a critical part of a user-facing Lucky app is CSS, but we currently don't provide anything except NormalizeCSS. Having a templating system where users could pick between Bootstrap, Bulma, Tailwind, etc _could_ be pretty cool, and again takes some of the burden off of our team from having to "declare a winner" as the framework default.

**Comment by paulcsmith** [comment]
Summary: I think we should make a lucky task in Authentic that generates most of the files, and outputs instructions for the bits it can't do easily (like adding to existing files, e.g. the `User` model columns). Later on I think we should have a way to install a templates with a lucky task `lucky template` 

This is a tricky one. Because we want to balance providing code that is tedious to do write yourself, without introducing too many options. I think if we provide things by default it should be easy to remove *or* useful in the majority of cases (i.e. very few people would want to remove it). In this case there are a number of files generated, columns added to the migration and model, etc. And not everyone wants/likes email confirmation. So it is hard to remove.

We could do the devise route but I am personally not a fan because it is super hard to customize and understand what's going on because so much is hidden. I like generating most of the code instead, like we do now with `lucky init`

So I'd say we should do one of these two things:

1. Add another option to the generator to generate locking/confirmation/etc. Not ideal because the generator will just grow and grow, and you can't add confirmation/locking/etc. later if you decide you want it.
2. Make an installer that adds templates in
  * Generates migration that adds confirmed_at, etc.
  * Generates actions/operations/pages
  * May need to output something for the stuff you have to add yourself ("Add column X to your model")
    * Ideally we could add the columns by finding where to put it inside `table do/end` but not sure how hard that'd be

I think the real question is *how* do we do this. We could probably do a V1 with just teeplate and a lucky task in Authentic that generates the files `lucky authentic.templates.email_confirmation`. I think this is a good start for MVP. We can even call the task in our Lucky CLI test suite to make sure it works once installed.

However, long-term I think we need our own Railsbyte type thing so we don't need users to install Bloat and Ruby, etc. I think we really need something like `lucky template link/to/template`. However I think that is a separate discussion (which I'll start in a moment)

**Reply by paulcsmith** [reply]
Discussion about adding ability to install templates https://github.com/luckyframework/lucky/discussions/1286
